src/Compute/v2/Api/Server.php

```diff
@@ -0,0 +1,204 @@
+<?php
+
+namespace OpenStack\Compute\v2\Api;
+
+final class Server
+{
+    public static function post()
+    {
+        return [
+            'path' => 'servers',
+            'method' => 'POST',
+            'jsonKey' => 'server',
+            'params' => [
+                'securityGroups' => [
+                    'type' => 'array',
+                    'items' => ['type' => 'object', 'items' => ['name' => ['type' => 'string']]],
+                    'sentAs' => 'security_groups',
+                ],
+                'userData' => ['type' => 'string', 'sentAs' => 'user_data'],
+                'availabilityZone' => ['type' => 'string', 'sentAs' => 'availability_zone'],
+                'imageId' => ['type' => 'string', 'required' => true, 'sentAs' => 'imageRef'],
+                'flavorId' => ['type' => 'string', 'required' => true, 'sentAs' => 'flavorRef'],
+                'networks' => [
+                    'type' => 'array',
+                    'items' => [
+                        'type' => 'object',
+                        'items' => [
+                            'uuid' => ['type' => 'string'],
+                            'port' => ['type' => 'string'],
+                        ]
+                    ]
+                ],
+                'name' => ['type' => 'string', 'required' => true],
+                'metadata' => ['type' => 'string'],
+                'personality' => ['type' => 'string'],
+                'blockDeviceMapping' => [
+                    'type' => 'array',
+                    'sentAs' => 'block_device_mapping_v2',
+                    'items' => [
+                        'type' => 'object',
+                        'items' => [
+                            'configDrive' => ['type' => 'string', 'sentAs' => 'config_drive'],
+                            'bootIndex' => ['type' => 'string', 'sentAs' => 'boot_index'],
+                            'deleteOnTermination' => ['type' => 'boolean', 'sentAs' => 'delete_on_termination'],
+                            'guestFormat' => ['type' => 'string', 'sentAs' => 'guest_format'],
+                            'destinationType' => ['type' => 'string', 'sentAs' => 'destination_type'],
+                            'sourceType' => ['type' => 'string', 'sentAs' => 'source_type'],
+                            'deviceName' => ['type' => 'string', 'sentAs' => 'device_name'],
+                        ]
+                    ],
+                ],
+            ]
+        ];
+    }
+
+    public static function getAll()
+    {
+        return [
```

Comment:
Yeah, totally. It's not 100% similar, but it uses some of the structure and definitions sparingly. When you say "use existing vocabularies", do you mean that we should replace our use of terminology with the WADL equivalents, effectively keeping our current implementation? Or do you think we should overhaul our implementation to actually use WADL/JSON schema?

The problem I have with WADL and any other vocabularies is that they get very verbose and offer us a bunch of functionality we will never need in the SDK. Also some of things we actually need have no equivalent in JSON schema or WADL (like aliasing for example). I want a more focused vocabulary that is lean and precise. An operation only has four keys:

`path` - the URL path
`method` - HTTP method
`jsonKey` - the top-level JSON key for request body (i.e. `{"server": {}}`)
`params` - an array of user parameters

Each parameter is nested, and has these keys:

`type` - the value type (string, integer, boolean)
`required` - whether or not the user must provide a value
`location` - either `url`, `json`, `body`, `header` or `query`
`sentAs` - the alias. So `contentType` will be sent as `Content-Type`
`path` - used for JSON bodies. it will insert the value into the JSON structure using an XPath-like definition.

Also, if we used WADL or JSON schemas, we'd probably need an external parsing library - which just adds to our dependency list - which I definitely don't want. I only think we need the above definitions, so IMHO it's not a major issue - we're just formalizing the stuff we'd want to store anyway.